xarray/tests/test_backends.py

```diff
@@ -1104,55 +1119,188 @@ def test_auto_chunk(self):
+        # encoding
         data = create_test_data()
         chunks = (5, 5)
         data['var2'].encoding.update({'chunks': chunks})
+
         with self.roundtrip(data) as actual:
             self.assertEqual(chunks, actual['var2'].encoding['chunks'])
+
+        # expect an error with non-integer chunks
         data['var2'].encoding.update({'chunks': (5, 4.5)})
-        with pytest.raises(ValueError):
+        with pytest.raises(TypeError):
             with self.roundtrip(data) as actual:
                 pass
 
+    def test_chunk_encoding_with_dask(self):
+        # These datasets DO have dask chunks. Need to check for various
+        # interactions between dask and zarr chunks
+        ds = xr.DataArray((np.arange(12)), dims='x', name='var1').to_dataset()
+
+        # - no encoding specified -
+        # zarr automatically gets chunk information from dask chunks
+        ds_chunk4 = ds.chunk({'x': 4})
+        with self.roundtrip(ds_chunk4) as actual:
+            self.assertEqual((4,), actual['var1'].encoding['chunks'])
+
+        # should fail if dask_chunks are irregular...
+        ds_chunk_irreg = ds.chunk({'x': (5, 4, 3)})
+        with pytest.raises(ValueError):
+            with self.roundtrip(ds_chunk_irreg) as actual:
+                pass
+
+        # ... except if the last chunk is smaller than the first
+        ds_chunk_irreg = ds.chunk({'x': (5, 5, 2)})
+        with self.roundtrip(ds_chunk_irreg) as actual:
+            self.assertEqual((5,), actual['var1'].encoding['chunks'])
+
+        # - encoding specified  -
+        # specify compatible encodings
+        for chunk_enc in 4, (4, ):
+            ds_chunk4['var1'].encoding.update({'chunks': chunk_enc})
+            with self.roundtrip(ds_chunk4) as actual:
+                self.assertEqual((4,), actual['var1'].encoding['chunks'])
+
+        # specify incompatible encoding
+        ds_chunk4['var1'].encoding.update({'chunks': (5, 5)})
+        with pytest.raises(ValueError):
+            with self.roundtrip(ds_chunk4) as actual:
+                pass
+
+        # TODO: remove this failure once syncronized overlapping writes are
+        # supported by xarray
+        ds_chunk4['var1'].encoding.update({'chunks': 5})
+        with pytest.raises(NotImplementedError):
+            with self.roundtrip(ds_chunk4) as actual:
+                pass
+
     def test_vectorized_indexing(self):
         self._test_vectorized_indexing(vindex_support=True)
 
+    def test_hidden_zarr_keys(self):
+        expected = create_test_data()
+        with self.create_store(mode='w') as store:
+            expected.dump_to_store(store)
+            zarr_group = store.ds
+
+            # check that the global hidden attribute is present
+            assert self.DIMENSION_KEY in zarr_group.attrs
+
+            # check that a variable hidden attribute is present and correct
+            # for some reason, one is a list and the other a tuple
```

Comment:
To answer: JSON only has a single array type, which maps to `list` in Python. In contrast, `dims` in xarray is always a `tuple`.